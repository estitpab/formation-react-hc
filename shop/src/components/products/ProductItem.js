import React from "react";
import AddToCartButton from "../cart/AddToCartButton";

const ProductItem = ({ product }) => {
  return (
    <div className="ProductItem col s3">
      <div className="card">
        <div className="card-image">
          <img alt="" src={product.image} />
          <AddToCartButton productId={product.id} />
        </div>
        <div className="card-content">
          <span className="card-title">
            {product.title} ({product.price} €)
          </span>
        </div>
      </div>
    </div>
  );
};

export default ProductItem;
