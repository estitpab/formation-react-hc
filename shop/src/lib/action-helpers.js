// generator (helper)
export const asyncActionCreatorFactory = (asyncFunction, prefix) => {
  // action creator
  return (...params) => {
    // action (intercepted by thunkMiddleware)
    return dispatch => {
      dispatch({
        type: `${prefix}_START`
      });
      return asyncFunction(...params)
        .then(cart =>
          dispatch({
            type: `${prefix}_SUCCESS`,
            payload: { result: cart }
          })
        )
        .catch(error =>
          dispatch({
            type: `${prefix}_ERROR`,
            error
          })
        );
    };
  };
};
