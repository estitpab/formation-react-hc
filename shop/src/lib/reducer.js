import { combineReducers } from "redux";
import * as app from "../components/app/app.reducer";
import * as cart from "../components/cart/cart.reducer";
// import * as products from '../components/products/products.reducer';

export const initialState = {
  app: app.initialState,
  cart: cart.initialState
};

export const reducer = combineReducers({
  app: app.reducer,
  cart: cart.reducer
});

/*
// List modules here: initialState and rootReducer will be computed from this list
const modules = {
  app,
  cart,
  //products,
};

// Compute exported initialState from reducers' initial states
export const initialState = Object.keys(modules).reduce((state, key) => {
  state[key] = modules[key].initialState; // eslint-disable-line no-param-reassign
  return state;
}, {});

// Combine small reducers to generate the global state and reducer
export const reducer = combineReducers(
  Object.keys(modules).reduce((reducers, key) => {
    reducers[key] = modules[key].reducer; // eslint-disable-line no-param-reassign
    return reducers;
  }, {}),
);
*/
