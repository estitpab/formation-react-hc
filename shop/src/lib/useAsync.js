import React from "react";
//import useDidMount from "./useDidMount";

// Declared globally to share reference between calls
const initialStatus = {
  loading: true,
  result: null,
  error: null
};

const useAsync = asyncFunction => {
  const [{ loading, result, error }, setStatus] = React.useState(initialStatus);

  React.useEffect(() => {
    // start
    setStatus(initialStatus); // reset status
    asyncFunction()
      .then(result => setStatus({ loading: false, result }))
      .catch(error => setStatus({ loading: false, error }));
    return () => {
      // TODO cancel request?
    };
  }, [asyncFunction]);

  return [error, loading, result];
};

export default useAsync;
