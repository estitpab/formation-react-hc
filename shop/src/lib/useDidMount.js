import React from "react";

const useDidMount = fn => {
  React.useEffect(fn, []);
};

export default useDidMount;
