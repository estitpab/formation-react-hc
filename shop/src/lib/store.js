import { createStore, applyMiddleware, compose } from "redux";
import { reducer, initialState as defaultInitialState } from "./reducer";
import { getCart } from "./api";
import {
  getCartStart,
  getCartSuccess,
  getCartError
} from "../components/cart/cart.actions";

const pendingActions = {};
const setPageLoading = store => next => action => {
  /* Exemple d'effet de bord: injecter dans l'action des données d'autres parties du store
  // store.getState()
  if (action.type === 'CART_…') {
    action.meta.categories = store.getState().app.categories
    next(action)
  }
  */

  if (action?.type?.match(/_START$/)) {
    pendingActions[action.type.replace(/_START$/, "")] = true;
  }
  if (action?.type?.match(/_(SUCCESS|ERROR)$/)) {
    delete pendingActions[action.type.replace(/_(SUCCESS|ERROR)$/, "")];
  }
  if (Object.keys(pendingActions).length > 0) {
    document.title = "(...) " + document.title.replace(/^\(\.\.\.\) /, "");
  } else {
    document.title = document.title.replace(/^\(\.\.\.\) /, "");
  }

  next(action);
};

// npm add redux-thunk
const thunkMiddleware = store => next => action => {
  if (typeof action === "function") {
    return action(store.dispatch, store.getState);
  }
  return next(action);
};

export const initStore = (initialState = defaultInitialState) => {
  const middlewares = [setPageLoading, thunkMiddleware];

  // Wrapper to enable Redux Devtools
  // https://github.com/zalmoxisus/redux-devtools-extension#usage
  const composeEnhancers =
    process.env.NODE_ENV !== "production" && // not in production
    typeof window === "object" &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
          // Specify extension’s options like name, actionsBlacklist, actionsCreators, serialize...
        })
      : compose;

  // Apply middlewares
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));

  // Initialize store with global reducer, global initial state, and enhanced middlewares
  return createStore(reducer, initialState, enhancer);
};
