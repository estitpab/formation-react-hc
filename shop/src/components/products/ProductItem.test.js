import React from "react";
import ProductItem from "./ProductItem";
import { render, fireEvent } from "@testing-library/react";
import { initStore } from "../../lib/store";
import { Provider } from "react-redux";

const mockFn = jest.fn();

jest.mock("../../lib/api", () => ({
  addToCart: async (productId, qty) => {
    mockFn(productId, qty);
  }
}));

test("ProductItem", () => {
  const product = {
    id: "id",
    image: "image",
    title: "title",
    price: 42
  };

  const store = initStore();

  const element = (
    <Provider store={store}>
      <ProductItem product={product} />
    </Provider>
  );
  const wrapper = render(element);
  // console.log(wrapper);

  expect(wrapper.container).toMatchSnapshot();

  const addShoppingCart = wrapper.getByText("add_shopping_cart");
  expect(addShoppingCart).toBeDefined();

  fireEvent.click(addShoppingCart);
  expect(mockFn).toHaveBeenCalled();
  expect(mockFn).toHaveBeenCalledWith("id", 1);
});
