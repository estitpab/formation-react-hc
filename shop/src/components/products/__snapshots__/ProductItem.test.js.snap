// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem 1`] = `
<div>
  <div
    class="ProductItem col s3"
  >
    <div
      class="card"
    >
      <div
        class="card-image"
      >
        <img
          alt=""
          src="image"
        />
        <button
          class="btn-floating btn-large halfway-fab waves-effect waves-light red"
        >
          <i
            class="material-icons"
          >
            add_shopping_cart
          </i>
        </button>
      </div>
      <div
        class="card-content"
      >
        <span
          class="card-title"
        >
          title
           (
          42
           €)
        </span>
      </div>
    </div>
  </div>
</div>
`;
